saleor/checkout/calculations.py

```diff
@@ -96,18 +109,187 @@ def checkout_line_total(
-        discounts or [],
+    return (
+        fetch_checkout_prices_if_expired(
+            checkout_info,
+            manager=manager,
+            lines=lines,
+            address=address,
+            discounts=discounts,
+        )
+        .lines.get(pk=checkout_line_info.line.pk)
+        .total_price
+    )
+
+
+def checkout_line_unit_price(
+    *,
+    manager: "PluginsManager",
+    checkout_info: "CheckoutInfo",
+    lines: Iterable["CheckoutLineInfo"],
+    checkout_line_info: "CheckoutLineInfo",
+    discounts: Optional[Iterable[DiscountInfo]] = None,
+) -> "TaxedMoney":
+    address = checkout_info.shipping_address or checkout_info.billing_address
+    return (
+        fetch_checkout_prices_if_expired(
+            checkout_info,
+            manager=manager,
+            lines=lines,
+            address=address,
+            discounts=discounts,
+        )
+        .lines.get(pk=checkout_line_info.line.pk)
+        .unit_price
+    )
+
+
+def fetch_checkout_prices_if_expired(
+    checkout_info: "CheckoutInfo",
+    manager: "PluginsManager",
+    lines: Iterable["CheckoutLineInfo"],
+    address: Optional["Address"] = None,
+    discounts: Optional[Iterable["DiscountInfo"]] = None,
+    force_update: bool = False,
+) -> "Checkout":
+    checkout = checkout_info.checkout
+
+    if not force_update and checkout.price_expiration < timezone.now():
+        return checkout
+
+    plugins_tax_data = _get_tax_data_from_plugins(
+        checkout, manager, checkout_info, lines, address, discounts
     )
-    return quantize_price(calculated_line_total, checkout_info.checkout.currency)
+
+    webhooks_tax_data = manager.get_taxes_for_checkout(checkout)
+
+    if webhooks_tax_data:
+        _apply_tax_data(checkout, webhooks_tax_data)
+    elif plugins_tax_data:
+        _apply_tax_data(checkout, plugins_tax_data)
+
+    return checkout
+
+
+def _apply_tax_data(checkout: "Checkout", tax_data: TaxData) -> None:
+    def QP(price):
+        return quantize_price(price, checkout.currency)
+
+    checkout.total_net_amount = QP(tax_data.total_net_amount)
+    checkout.total_gross_amount = QP(tax_data.total_gross_amount)
+
+    checkout.subtotal_net_amount = QP(tax_data.subtotal_net_amount)
+    checkout.subtotal_gross_amount = QP(tax_data.subtotal_gross_amount)
+
+    checkout.shipping_price_net_amount = QP(tax_data.shipping_price_net_amount)
+    checkout.shipping_price_gross_amount = QP(tax_data.shipping_price_gross_amount)
+
+    checkout.price_expiration += settings.CHECKOUT_PRICES_TTL
+    checkout.save(
+        update_fields=[
+            "total_net_amount",
+            "total_gross_amount",
+            "subtotal_net_amount",
+            "subtotal_gross_amount",
+            "shipping_price_net_amount",
+            "shipping_price_gross_amount",
+            "price_expiration",
+        ]
+    )
+
+    checkout_lines = checkout.lines.all()
+
+    for (checkout_line, tax_line_data) in zip(checkout_lines, tax_data.lines):
```

Comment:
You shouldn't relay on the order of the returned lines. These data (checkout_lines) are sending via webhook sync, we don't know how the future apps will process it, but we can't assume that we will receive these data in the same order as we have previously.